[data] refactor download expression to use inheritance from `AbstractOneToOne`

omatthew98 opened the pull request:
## Why are these changes needed?
`Download` logical operator isn't really a map, we should use `AbstractOneToOne` as the base class instead.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

omatthew98 committed:
refactor

Signed-off-by: Matthew Owen <mowen@anyscale.com>

bveeramani reviewed (approved):
LGTM assuming no regression

omatthew98 commented on python/ray/data/_internal/logical/operators/one_to_one_operator.py:
```diff
@@ -83,3 +83,37 @@ def _input_files(self):
+        self._output_bytes_column_name = output_bytes_column_name
+        self._ray_remote_args = ray_remote_args or {}
+
+    def can_modify_num_rows(self) -> bool:
+        return False
```
I think it will depend on our implementation of the decode API. Right now every row will return the bytes of the uri, if we allow the decode function to return multiple rows we will need to handle that separately. For now, I think it is most correct to leave this as is.

omatthew98 committed:
Update python/ray/data/_internal/logical/operators/one_to_one_operator.py

Co-authored-by: Balaji Veeramani <bveeramani@berkeley.edu>
Signed-off-by: Matthew Owen <omatthew98@berkeley.edu>

omatthew98 commented:
> LGTM assuming no regression

Release test results:
* Fixed size: 219.08s (build [here](https://buildkite.com/ray-project/release/builds/56926#01991c06-6367-42b4-afc6-1d4b80513260))
* Autoscaling: 488.46 (build [here](https://buildkite.com/ray-project/release/builds/56926#01991c06-6368-4777-97f6-5925e5fe5c76))
[Previous PR/benchmark](https://github.com/ray-project/ray/pull/56091#issuecomment-3246959075):
Fixed size: 229.85s ([build](https://buildkite.com/ray-project/release/builds/56244#01990c6c-0974-4b38-a2a9-acdbf0f84be4))
Autoscaling: 500.54s ([build](https://buildkite.com/ray-project/release/builds/56244#01990c6c-0975-42d4-a13c-4c4e6d2e55af))
